piker/brokers/kraken.py

```diff
@@ -764,6 +764,23 @@ async def subscribe(ws: wsproto.WSConnection, token: str):
                         
                         await ems_stream.send(trade_msg.dict())
 
+                        filled_msg = BrokerdStatus(
+                            reqid=trade.reqid,
+                            time_ns=time.time_ns(),
+
+                            account='kraken.spot',
+                            status='filled',
+                            filled=float(trade.size),
+                            reason='Order filled by kraken',
+                            # remaining='' # TODO: not sure what to do here.
```

Comment:
yeah it's because you haven't seen the general case.
basically a fill can *use up* (aka subtract from) the current order's "size".

So say you put in an order for 1000 of something but only 400 get's filled, then shortly later another 600; that would terminate the order's lifetime - it would be completely. However multiple fills can be used to "consume" up the order. More technically, and order can be completed over multiple *clearing prices*.

Reply:
made https://github.com/pikers/piker/issues/296 as follow up.

Reply:
Yeah so the main thing you're missing here is pretty simple:

- you've got your `Trade(BaseModel)` type that is used to parse the details from kraken
- you've also got a message you **shouldn't be sending** which is the first one with the `status=executed`. The reason you didn't catch this is bc [we're only logging that message](https://github.com/pikers/piker/blob/master/piker/clearing/_ems.py#L705) instead of making sure it's in the correct set; we should make a bug report for this.
- you've got 2 msgs you need to send to the ems:
  - a `BrokerdStatus` with status 'filled' which indicates that (some of) the order was filled and there is [logic in the ems's relay loop](https://github.com/pikers/piker/blob/master/piker/clearing/_ems.py#L689) which checks the `.remaining` field on that message. [If the remaining field is `bool` null (`0` or `None`) then the ems takes care of sending the `broker_executed` msg to the client side](https://github.com/pikers/piker/blob/master/piker/clearing/_ems.py#L693). This message doesn't have to declare the fill/execution details if the backend does not want to but *it can* in the following additional message
  - a `BrokerdFill` message *can* be sent if you expect multiple fills to eventually fulfill the *execution* (i.e. all clears of) the logical *order*. These msgs [are also relayed to the client](https://github.com/pikers/piker/blob/master/piker/clearing/_ems.py#L712) which allows it to draw multiple annotations on the chart (triangles) that should all the individual fills that eventually satisfy the order's execution.

TL;DR:
- don't send the first msg, it's noto valid but we need to actually enforce that
- the `filled` status should decrement it's `.remaining` on every `Trade` you get from the kraken end
- `BrokerdFill` msgs should be sent 1-to-1 for every `Trade` msg you get from kraken so that the UI can show the multiple clears on chart.

Reply:
made https://github.com/pikers/piker/issues/297 to address why you originally were sending the wrong `.status`.

Reply:
thanks, this sounds great. it is also much more clear now. I like the idea of enforcing not making the mistake I did.